### PR TITLE
Add filter by environment and tags

### PR DESCRIPTION
**Pull request by penja**
PR for this isse https://github.com/Scalr/scalr-vscode/issues/10

### CODE REVIEW

File: src/providers/workspaceProvider.ts [hunk 1]
```diff
@@ -120,9 +148,112 @@ export class WorkspaceTreeDataProvider implements vscode.TreeDataProvider<vscode
         });
     }
 
-    dispose() {
-        //
+    private async chooseFilterOrClear() {
+        const filterType = await vscode.window.showQuickPick(Object.values(WorkspaceFilter), {
+            placeHolder: 'Choose a filter type or clear all filters',
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:153]:
```suggestion
            placeHolder: 'Select one of the filters below',
```

File: src/providers/workspaceProvider.ts [hunk 2]
```diff
@@ -120,9 +148,112 @@ export class WorkspaceTreeDataProvider implements vscode.TreeDataProvider<vscode
...
+    private async enterFilterValue(filterType: string) {
+        const delay = 500;
+        let typingTimer: NodeJS.Timeout | null = null;
+
+        switch (filterType) {
+            case WorkspaceFilter.environment: {
+                const environmentsQuickPicks = vscode.window.createQuickPick();
+                environmentsQuickPicks.items = await this.getEnvironmentQuickPick();
+                environmentsQuickPicks.title = 'Select one or more environments';
+                environmentsQuickPicks.placeholder = 'Type to search by query';
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:173]:
```suggestion
                environmentsQuickPicks.placeholder = 'Enter a workspace name or ID';
```

File: src/providers/workspaceProvider.ts [hunk 3]
```diff
@@ -120,9 +148,112 @@ export class WorkspaceTreeDataProvider implements vscode.TreeDataProvider<vscode
...
+                    environmentsQuickPicks.hide();
+                });
+
+                environmentsQuickPicks.show();
+                break;
+            }
+            case WorkspaceFilter.query: {
+                const currentQuery = (this.filters.get('query') || '') as string;
+                const query = await vscode.window.showInputBox({
+                    placeHolder: currentQuery || 'Type to search by query',
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:210]:
```suggestion
                    placeHolder: currentQuery || 'Enter a workspace name or ID',
```

File: src/providers/workspaceProvider.ts [hunk 4]
```diff
@@ -120,9 +148,112 @@ export class WorkspaceTreeDataProvider implements vscode.TreeDataProvider<vscode
...
+                    prompt: 'Filter by query',
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:211]:
```suggestion
                    prompt: 'Filtering by workspace name or ID',
```

File: src/providers/workspaceProvider.ts [hunk 5]
```diff
@@ -120,9 +148,112 @@ export class WorkspaceTreeDataProvider implements vscode.TreeDataProvider<vscode
...
+    private async getEnvironmentQuickPick(query?: string): Promise<QuickPickItem[]> {
+        const { data, error } = await listEnvironments({
+            query: {
+                fields: { environments: 'name' },
+                query: query,
+            },
+        });
+
+        if (error || !data) {
+            vscode.window.showErrorMessage('Failed to fetch environments' + error);
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:242]:
```suggestion
            vscode.window.showErrorMessage('Unable to get environments: ' + error);
```

File: src/providers/workspaceProvider.ts [hunk 6]
```diff
@@ -173,6 +304,33 @@ class LoadMoreItem extends vscode.TreeItem {
         this.command = {
             command: 'workspace.loadMore',
             title: 'Show more',
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:306]:
```suggestion
            title: 'Next page',
```

File: src/providers/workspaceProvider.ts [hunk 7]
```diff
@@ -173,6 +304,33 @@ class LoadMoreItem extends vscode.TreeItem {
...
+            tooltip: 'Load more workspaces',
         };
     }
 }
+
+class FilterInfoItem extends vscode.TreeItem {
+    constructor(private filters: Map<WorkspaceFilterApiType, string[] | string>) {
+        super('Applied Filters', vscode.TreeItemCollapsibleState.None);
+        this.description = Array.from(filters.entries())
+            .map(([key, value]) => `${key}: ${value}`)
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:316]: let's show names instead of IDs.

File: src/providers/workspaceProvider.ts [hunk 8]
```diff
@@ -173,6 +304,33 @@ class LoadMoreItem extends vscode.TreeItem {
...
+    constructor(
+        public label: string,
+        public id: string
+    ) {}
+}
+
+enum WorkspaceFilter {
+    //important the key value must be the same as the filter key in the API
+    environment = 'By Environments',
+    query = 'By Query',
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:333]:
```suggestion
    query = 'By workspace name of ID',
```

File: src/providers/workspaceProvider.ts [hunk 9]
```diff
@@ -173,6 +304,33 @@ class LoadMoreItem extends vscode.TreeItem {
...
+class QuickPickItem implements vscode.QuickPickItem {
+    constructor(
+        public label: string,
+        public id: string
+    ) {}
+}
+
+enum WorkspaceFilter {
+    //important the key value must be the same as the filter key in the API
+    environment = 'By Environments',
```
Comment (emocharnik) [inline src/providers/workspaceProvider.ts:332]:
```suggestion
    environment = 'By environments',
```

**Review by emocharnik — COMMENTED** [review]
Updated wording a bit.

File: src/providers/workspaceProvider.ts (hunk 7 above)
Comment (penja) [reply src/providers/workspaceProvider.ts:316]: Ok

**Pull request merged**